crates/kitsune_p2p/kitsune_p2p/src/gossip/simple_bloom.rs

```diff
@@ -0,0 +1,548 @@
+}
+
+pub(crate) enum HowToConnect {
+    Con(Tx2ConHnd<wire::Wire>),
+    Url(TxUrl),
+}
+
+pub(crate) struct SimpleBloomModInner {
+    tuning_params: KitsuneP2pTuningParams,
+    space: Arc<KitsuneSpace>,
+    ep_hnd: Tx2EpHnd<wire::Wire>,
+    evt_sender: futures::channel::mpsc::Sender<event::KitsuneP2pEvent>,
+
+    local_agents: HashSet<Arc<KitsuneAgent>>,
+    local_bloom: BloomFilter,
+    local_data_map: DataMap,
+    local_key_set: KeySet,
+
+    remote_metrics: HashMap<Tx2Cert, NodeInfo>,
+
+    last_initiate_check: std::time::Instant,
+    initiate_tgt: Option<Tx2Cert>,
+
+    incoming: Vec<(Tx2ConHnd<wire::Wire>, GossipWire)>,
+
+    last_outgoing: std::time::Instant,
+    send_interval_ms: u64,
+    outgoing: Vec<(Tx2Cert, HowToConnect, GossipWire)>,
+}
+
+impl SimpleBloomModInner {
+    pub fn new(
+        tuning_params: KitsuneP2pTuningParams,
+        space: Arc<KitsuneSpace>,
+        ep_hnd: Tx2EpHnd<wire::Wire>,
+        evt_sender: futures::channel::mpsc::Sender<event::KitsuneP2pEvent>,
+    ) -> Self {
+        let send_interval_ms: u64 = (
+            // !*)&^$# cargo fmt...
+            16384.0    // max bytes in a gossip msg
+            * 8.0      // bits per byte
+            * 1000.0   // milliseconds
+            / 1024.0   // kbps
+            / 1024.0   // mbps
+            / tuning_params.gossip_output_target_mbps
+        ) as u64;
+
+        // pick an old instant for initialization
+        let old = std::time::Instant::now()
+            .checked_sub(std::time::Duration::from_secs(60 * 24))
+            .unwrap();
+
+        Self {
+            tuning_params,
+            space,
+            ep_hnd,
+            evt_sender,
+
+            local_agents: HashSet::new(),
+            local_bloom: bloomfilter::Bloom::new(1, 1),
+            local_data_map: HashMap::new(),
+            local_key_set: HashSet::new(),
+
+            remote_metrics: HashMap::new(),
+
+            last_initiate_check: old,
+            initiate_tgt: None,
+
+            incoming: Vec::new(),
+
+            last_outgoing: old,
+            send_interval_ms,
+            outgoing: Vec::new(),
+        }
+    }
+}
+
+enum GossipIterationResult {
+    Close,
+    Good,
+}
+
+enum CheckResult {
+    Close,
+    NotReady,
+    SyncAndInitiate,
+    SkipSyncAndInitiate,
+}
+
+struct SimpleBloomMod(Share<SimpleBloomModInner>);
+
+impl SimpleBloomMod {
+    pub fn new(
+        tuning_params: KitsuneP2pTuningParams,
+        space: Arc<KitsuneSpace>,
+        ep_hnd: Tx2EpHnd<wire::Wire>,
+        evt_sender: futures::channel::mpsc::Sender<event::KitsuneP2pEvent>,
+    ) -> Arc<Self> {
+        let inner = SimpleBloomModInner::new(tuning_params, space, ep_hnd, evt_sender);
+
+        let send_interval_ms = inner.send_interval_ms;
+
+        let this = Arc::new(Self(Share::new(inner)));
+
+        // this value needs to be somewhat frequent to support send timing
+        let loop_check_interval_ms = std::cmp::max(send_interval_ms / 3, 100);
```

Comment:
Is this going to cause problems if the ratio between send and check is closer to 1 than 3? (due to the `max`?)

Reply:
No, it'll increase the variability in how often we make outgoing gossip requests, but we need the minimum time there to avoid thrashing the database... (it always throws me off that you have to use the `max` function to achieve a minimum value...)